File: src/main/java/com/study/csdojo/Problem5_OneEditAway.py
def is_one_away(s1, s2):
    if abs(len(s1) - len(s2)) > 1:
        return False
    elif len(s1) == len(s2):
        return is_one_away_same_length(s1, s2)
    elif len(s1) > len(s2):
        return is_one_away_diff_length(s1, s2)
    else:
        return is_one_away_diff_length(s2, s1)

def is_one_away_same_length(s1, s2):
    diff_count = 0
    for i in range(len(s1)):
        if s1[i] != s2[i]:
            diff_count += 1
            if diff_count > 1:
                return False
    return True

def is_one_away_diff_length(s1, s2):
    diff_count = 0
    i = 0
    while i < len(s2):
        if s1[i + diff_count] == s2[i]:
            i += 1
        else:
            diff_count += 1
            if diff_count > 1:
                return False
    return True

if __name__ == '__main__':
    print("%s should return False" % (is_one_away("abcde", "abme")))
    print("%s should return True" % (is_one_away("abcde", "abcd")))
    print("%s should return True" % (is_one_away("abde", "abcde")))
    print("%s should return True" % (is_one_away("a", "a")))
    print("%s should return True" % (is_one_away("abcdef", "abqdef")))
    print("%s should return True" % (is_one_away("abcdef", "abccef")))
    print("%s should return True" % (is_one_away("abcdef", "abcde")))
    print("%s should return False" % (is_one_away("aaa", "abc")))
    print("%s should return False" % (is_one_away("abcde", "abc")))
    print("%s should return False" % (is_one_away("abc", "abcde")))
    print("%s should return False" % (is_one_away("abc", "bcc")))

File: src/main/java/com/study/csdojo/Problem5_OneEditAway.java
package com.study.csdojo;

public class Problem5_OneEditAway {

    public Boolean isOneAway(String s1, String s2) {
        if (Math.abs(s1.length() - s2.length()) > 1) {
            return false;
        } else if (s1.length() == s2.length()) {
            return isOneAwaySameLength(s1, s2);
        } else if (s1.length() > s2.length()) {
            return isOneAwayDiffLength(s1, s2);
        } else {
            return isOneAwayDiffLength(s2, s1);
        }
    }

    private Boolean isOneAwaySameLength(String s1, String s2) {
        int diffCount = 0;
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                diffCount++;
                if (diffCount > 1) {
                    return false;
                }
            }
        }
        return true;
    }

    private Boolean isOneAwayDiffLength(String s1, String s2) {
        int diffCount = 0;
        int i = 0;
        while (i < s2.length()) {
            if (s1.charAt(i + diffCount) != s2.charAt(i)) {
                diffCount++;
                if (diffCount > 1) {
                    return false;
                }
            } else {
                i++;
            }
        }
        return true;
    }

}
